contracts/PSM/UpperPegArbitrager.sol

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.19;
+
+import { Ownable2Step } from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import { IERC3156FlashBorrower } from "@openzeppelin/contracts/interfaces/IERC3156FlashBorrower.sol";
+import { IERC3156FlashLender } from "@openzeppelin/contracts/interfaces/IERC3156FlashLender.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { OneInchV5AMM } from "../AMMs/OneInchV5AMM.sol";
+import { IPSM } from "./IPSM.sol";
+import { IPSMFeeCalculator } from "./IPSMFeeCalculator.sol";
+import { IUpperPegArbitrager } from "./IUpperPegArbitrager.sol";
+
+contract UpperPegArbitrager is IUpperPegArbitrager, IERC3156FlashBorrower, OneInchV5AMM, Ownable2Step {
+    using SafeERC20 for IERC20;
+
+    enum Action {
+        NORMAL,
+        OTHER
+    }
+
+    IERC3156FlashLender public override lender;
+    IERC20 public override borrowToken;
+    IPSM public override psm;
+    IPSMFeeCalculator public override feeCalculator;
+
+    bytes private _dataForOneInchSwap;
+
+    constructor(
+        IERC3156FlashLender lender_,
+        IERC20 borrowToken_,
+        IPSM psm_,
+        address _aggregationRouter
+    )
+        OneInchV5AMM(_aggregationRouter)
+    {
+        if (address(lender_) == address(0)) {
+            revert ZeroInputProvided();
+        }
+        if (address(borrowToken_) == address(0)) {
+            revert ZeroInputProvided();
+        }
+        if (address(psm_) == address(0)) {
+            revert ZeroInputProvided();
+        }
+
+        lender = lender_;
+        borrowToken = borrowToken_;
+        psm = psm_;
+        feeCalculator = psm.feeCalculator();
+
+        borrowToken_.approve(address(psm), type(uint256).max);
+    }
+
+    /// @dev ERC-3156 Flash loan callback
+    function onFlashLoan(
+        address initiator,
+        address token,
+        uint256 amount,
+        uint256 fee,
+        bytes calldata data
+    )
+        external
+        override
+        returns (bytes32)
+    {
+        if (msg.sender != address(lender)) {
+            revert UntrustedLender(msg.sender);
+        }
+        if (initiator != address(this)) {
+            revert UntrustedLoanInitiator(initiator);
+        }
+
+        (Action action) = abi.decode(data, (Action));
+        if (action == Action.NORMAL) {
+            if (IERC20(token).balanceOf(address(this)) < amount) {
+                revert InsufficientBalanceAfterFlashLoan();
+            }
+
+            // Buy R from PSM
+            _buyR(amount + fee);
```

Comment:
why `+ fee`?

Reply:
I think this would brek if fee for DAI flash loan is `> 0`

Reply:
Well, and yes and no :) I used an example from Maker (and there is this `amount + fee` part) because we'll use that and the code for a fee is next
```
function flashFee(
        address token,
        uint256 amount
    ) external override view returns (uint256) {
        amount;
        require(token == address(dai), "DssFlash/token-unsupported");

        return 0;
    }
    ```
    So, it is always zero, but you are right, generally speaking, if we go with some other flash loan (that also support this interface), we'll be REKT. 
    
    I'll fix this in a few minutes.